script/rewrite_policy_publication_history.rb

```diff
@@ -4,27 +4,44 @@
 # NOTE: This script is NOT idempotent and so should only be run once!
 require 'csv'
 
-mapping_csv_path = Rails.root+"lib/tasks/election/2015-04-20-13-42-38-policy_paper_creation_output.csv"
 logger = Logger.new(STDOUT)
 
-# Make sure no publications have an open edition as we can't handle those
-CSV.parse(File.open(mapping_csv_path).read, headers: true).each do |row|
-  publication = Publication.find_by_id(row["policy_paper_id"])
-  next unless publication
+def with_publications_and_policies(&block)
```

Comment:
The first loop is a sanity check to make sure that the data is in a good state to actually perform the backfill so we can fail early.
The third loop performs the bulk of the work inside a transaction
The fourth loop results in jobs being queued to reindex all the backfilled publications, and needs to happen outside of the transaction loop of loop three, otherwise the workers will not be able to find the records (they won't be committed until the transaction ends)

I guess the checks in loop two could be done in loop one, but then the code would need to be much more complicated to get the same logging output (i.e. record the output for each of the separate sections before logging them at the end of the loop). For example:

```
I, [2015-05-27T12:06:27.675081 #13869]  INFO -- : Checking for publications with open editions on them...
I, [2015-05-27T12:06:30.385631 #13869]  INFO -- : Checking for publications with suspect first_published_at dates...
W, [2015-05-27T12:06:30.816717 #13869]  WARN -- : Publication 489791 to be repaired: Frst published is 2012-11-17 06:00:00 +0000; policy has 2012-11-07 17:07:37 +0000
W, [2015-05-27T12:06:31.001134 #13869]  WARN -- : Publication 489815 to be repaired: Frst published is 2012-12-12 06:00:00 +0000; policy has 2012-11-01 09:50:18 +0000
W, [2015-05-27T12:06:31.232657 #13869]  WARN -- : Publication 489844 to be repaired: Frst published is 2012-12-12 06:00:00 +0000; policy has 2012-10-03 16:02:54 +0100
W, [2015-05-27T12:06:31.297811 #13869]  WARN -- : Publication 489852 to be repaired: Frst published is 2015-05-07 06:00:00 +0100; policy has 2012-10-12 14:25:26 +0100
W, [2015-05-27T12:06:31.371304 #13869]  WARN -- : Publication 489861 to be repaired: Frst published is 2013-04-10 06:00:00 +0100; policy has 2013-04-09 15:00:40 +0100
W, [2015-05-27T12:06:32.003934 #13869]  WARN -- : Publication 489933 to be repaired: Frst published is 2015-05-07 06:00:00 +0100; policy has 2013-01-22 12:00:08 +0000
W, [2015-05-27T12:06:32.018290 #13869]  WARN -- : Publication 489935 to be repaired: Frst published is 2013-01-22 06:00:00 +0000; policy has 2012-10-12 14:17:48 +0100
I, [2015-05-27T12:06:32.208101 #13869]  INFO -- : Back-filling publication histories...
I, [2015-05-27T12:06:32.298551 #13869]  INFO -- : Backfilling history for: (489731) 2010 to
```

As you can see from the timings, those first two loops don't take particularly long. This is a one-off script that will only ever be run the once. I think we can live with the inefficiency of looping through the CSV an extra time.